### PR TITLE
Pop deleted assets' metadata instead of del'ing

### LINKED ISSUE #177: get_deleted: KeyError
possibly side effect of #175 fix (would have been nice if was accompanied with a unittest where all data is dropped first I guess)

```
(base) dandi@drogon:/mnt/backup/dandi/dandisets$ time flock -E 0 -e -n /home/dandi/.run/backup2datalad-cron.lock tools/backups2datalad-update-cron 000026
>> python -m tools.backups2datalad -l WARNING -J 5 --target . update-from-backup --zarr-target /mnt/backup/dandi/dandizarrs --backup-remote dandi-dandisets-dropbox --zarr-backup-remote dandi-dandizarrs-dropbox --gh-org dandisets --zarr-gh-org dandizarrs -e '000108$' 000026
2022-06-02T11:22:40-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/9ff64ef5-de9a-4318-93a9-9cc350c33865/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:23:43-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/8d9cc3c0-85f2-4a84-a768-15afc5167883/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:24:47-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/d4a79b88-07f8-49c7-91ff-d0d8cd237420/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:25:49-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/40a7c85e-26fc-4cc5-aba2-9fc6c48ffb94/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:26:51-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/5b2e0854-2954-4cf4-b97e-a6b3b1b177a1/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:27:53-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/dandisets/000026/versions/draft/assets/?order=created&page=304 in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:28:55-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/101f8122-728b-45b5-a3bd-c2ebec87cd9d/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:29:57-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/771c0559-0bc1-4c70-a32c-38c8a38b0b36/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:30:58-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/6ddd8100-ec41-40a4-8797-3bebde0e939f/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:31:56-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/62476044-d807-4e66-acde-e9b0f2b97562/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:32:56-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/f9d56016-37d2-4525-95b8-539a7469d862/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:33:56-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/a3f4edb9-2c9b-4353-acca-f16b325e9bfd/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:34:55-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/c377fd02-e3cc-4e27-9609-c7d6c9129a25/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:35:53-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/f5fa51d7-032e-4396-8226-fa2319412d2d/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:36:54-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/02003a93-ef3b-487f-9cfc-2b6edd4b8295/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:37:50-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/9c094585-6623-4922-8b81-a4ac77adceca/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:38:49-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/363e80c5-ea49-40b8-8215-4440ed851e9e/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:39:59-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/dandisets/000026/versions/draft/assets/?order=created&page=831 in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:41:12-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/bb1db0da-4c41-42b2-b468-04ea7ef2e816/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:42:22-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/d3bb47ee-6fb2-45ae-912d-f3a23f44b7ed/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:43:34-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/e35baffd-31c9-486e-8a28-d12eb23db6b0/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:44:45-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/2c2052e3-6186-4739-8f74-552d53f0f664/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:45:55-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/480ed302-95bd-422b-8c86-b25a75b518fb/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:47:05-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/6762cad9-d61d-4749-a061-34b0134fc501/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:48:15-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/71decc86-5376-4cdc-bcae-e8475ff04617/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:49:24-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/68288498-d7cd-454a-ad27-04c10738c521/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:50:32-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/ce853dad-e3a4-4dfd-ab3c-0214b8fe7bc0/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:51:41-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/e3af9704-9a99-4ff4-9418-f65b6ad10865/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:52:47-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/aa712df8-8f5f-4981-b9a2-5a8f57dd7dff/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:53:54-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/a5e4c7ba-d90e-4728-a4d7-920eca1e7560/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:55:00-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/fd40be01-820a-4566-87a3-ebd574685405/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:56:07-0400 [WARNING ] backups2datalad Retrying GET request to https://api.dandiarchive.org/api/assets/bf89e9b8-93a4-45b5-806e-0ff00409c83d/ in 1.000000 seconds as it raised ReadTimeout:
2022-06-02T11:56:35-0400 [ERROR   ] backups2datalad Operation failed with exception:
Traceback (most recent call last):
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/util.py", line 308, in dandi_logging
    yield logfile
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/datasetter.py", line 145, in sync_dataset
    syncer.prune_deleted()
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/syncer.py", line 49, in prune_deleted
    for asset_path in self.tracker.get_deleted(self.config):
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/util.py", line 255, in get_deleted
    del self.asset_metadata[apath]
KeyError: 'rawdata/sub_I55/ref_I55.btf'
Traceback (most recent call last):
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/__main__.py", line 327, in <module>
    main()
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/dandi/miniconda3/envs/dandisets/lib/python3.8/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/__main__.py", line 161, in update_from_backup
    datasetter.update_from_backup(dandisets, exclude=exclude)
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/datasetter.py", line 77, in update_from_backup
    changed = self.sync_dataset(d, ds)
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/datasetter.py", line 145, in sync_dataset
    syncer.prune_deleted()
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/syncer.py", line 49, in prune_deleted
    for asset_path in self.tracker.get_deleted(self.config):
  File "/mnt/backup/dandi/dandisets/tools/backups2datalad/util.py", line 255, in get_deleted
    del self.asset_metadata[apath]
KeyError: 'rawdata/sub_I55/ref_I55.btf'

real    35m10.752s
user    3m47.272s
sys     0m12.196s
```

FWIW, #175 seems to be achieving what was needed to fix #174 since I see a good number of deleted files (didn't check for specific target one):

```shell
(base) dandi@drogon:/mnt/backup/dandi/dandisets$ git -C 000026 status | nl
     1  On branch draft
     2  Your branch is up to date with 'github/draft'.

     3  Changes to be committed:
     4    (use "git restore --staged <file>..." to unstage)
     5          deleted:    derivatives/sub-I38/ses-SPIM/micr/manual-annotation/sub-I38_ses-SPIM_sample-BrocaAreaS01_stain-NeuN_SPIM_pos-115-8770-4382.marker
...
    99          deleted:    rawdata/sub-I46/ses-SPIM/sub-I46_ses-SPIM_sample-MotorCortexS08_stain-Calretinin_chunk-08_SPIM.ome.tiff
   100          deleted:    rawdata/sub-I46/ses-SPIM/sub-I46_ses-SPIM_sample-MotorCortexS08_stain-Nuclei_chunk-04_SPIM.ome.tiff
```

### PR DESCRIPTION
**Pull request by jwodder**
Fixes #177.

### CODE REVIEW

File: tools/backups2datalad/util.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ def get_deleted(self, config: Config) -> Iterator[str]:
         """
         for apath in self.local_assets:
             if config.match_asset(apath):
-                del self.asset_metadata[apath]
+                self.asset_metadata.pop(apath)
```
Comment (yarikoptic) [inline tools/backups2datalad/util.py]: why does it matter since I thought that `del ` on `dict[item]` would do the same as the `.pop` here?

Comment (jwodder) [reply tools/backups2datalad/util.py]: Fixed.

File: tools/backups2datalad/util.py [hunk 2]
```diff
@@ -252,7 +252,7 @@ def get_deleted(self, config: Config) -> Iterator[str]:
         """
         for apath in self.local_assets:
             if config.match_asset(apath):
-                del self.asset_metadata[apath]
+                self.asset_metadata.pop(apath, None)
```
Comment (yarikoptic) [inline tools/backups2datalad/util.py:255]: ah, you mean to not fail if `apath` isn't there in the dict? gotcha now.

Comment (yarikoptic) [reply tools/backups2datalad/util.py:255]: but I wish there was a test which would cover this part of the logic !

**Comment by yarikoptic** [comment]
thanks! Let's try it out

File: tools/backups2datalad/util.py (hunk 2 above)
Comment (jwodder) [reply tools/backups2datalad/util.py:255]: I'm not entirely sure what would cover it.  As far as I can tell, the only way `del` should have failed would be if an asset was absent from `assets.json` but present in the dataset, and I don't think that should happen.

Comment (yarikoptic) [reply tools/backups2datalad/util.py:255]: well -- our case is that broken symlink is present, but it is not listed among `.dandi/assets.json` and not listed in the output from API server.

**Pull request merged**